midend/flattenHeaders.cpp

```diff
@@ -0,0 +1,130 @@
+/*
+    Copyright 2018 MNK Consulting, LLC.
+    http://mnkcg.com
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "flattenHeaders.h"
+
+namespace P4 {
+
+void HeaderTypeReplacement::flatten(const P4::TypeMap* typeMap,
+                                    cstring prefix,
+                                    const IR::Type* type,
+                                    IR::IndexedVector<IR::StructField> *fields) {
+    if (auto st = type->to<IR::Type_StructLike>()) {
+        for (auto f : st->fields) {
+            auto ft = typeMap->getType(f, true);
+            flatten(typeMap, prefix + "." + f->name, ft, fields);
+        }
+        return;
+    }
+    cstring fieldName = prefix.replace(".", "_") +
+                        cstring::to_cstring(fieldNameRemap.size());
+    fieldNameRemap.emplace(prefix, fieldName);
+    fields->push_back(new IR::StructField(IR::ID(fieldName), type->getP4Type()));
```

Comment:
May be useful to add a `@name` annotation preserving the original name of the field - in case that it shows up in some error messages. Not sure how useful this will be, but it can't be harmful.

Reply:
Probably that should be the full nested name including dots.

Reply:
Ok.